### ISSUE TITLE
WinClosed hook does not trigger when quitting

### ISSUE THREAD

**Issue by Screwtapello** [mawww/kakoune#2830]
### Steps 

Launch Kakoune with a hook that writes to a file:

```
$ kak -n -e 'hook global WinClose .* %{ nop %sh{ echo closed >> /tmp/win-closed } }'
```

...then immediately exit with `:q`.

### Outcome 

```
 $ cat /tmp/win-closed
cat: /tmp/win-closed: No such file or directory
```

### Expected 

```
 $ cat /tmp/win-closed
closed
```

### Notes

If you run `:db` before `:q`, the file is created as expected.

I can reproduce this problem with the current master, a32da9d.

My use-case for this feature is the [state-save plugin](https://gitlab.com/Screwtapello/kakoune-state-save), which records `%val{selections_desc}` when a particular file is closed, and reloads the selection (with `:select`) the next time that file is opened.

This definitely used to work when I first developed the state-save plugin.

**Comment by andreyorst** [comment]
I want something similar for my [tagbar.kak](https://github.com/andreyorst/tagbar.kak). When user exits client check if other clients left, and if so, check if other clients are `tagbarclient` and kill them. I suppose that [client hooks](https://github.com/mawww/kakoune/issues/2765) are needed for that, but I was initially trying to achieve it with `WinClosed` hook too.

**Comment by mawww** [comment]
I pushed a fix for that, but I think ultimately we should get rid of WinClose, which runs at unexpected times due to how window lifetime is managed. I would not be surprised if there was ways to crash Kakoune with say creating new buffers in a WinClose hook that would run while we are destroying the buffer manager.

**Comment by Screwtapello** [comment]
That makes sense, but is there a better hook for "the user has finished with this window"? There's `BufClose` of course, but since I want to record the selection state that's no good.

**Comment by andreyorst** [comment]
> I would not be surprised if there was ways to crash Kakoune with say creating new buffers in a WinClose hook that would run while we are destroying the buffer manager.

Unfortunately yes. `tagbar.kak` tryes its best to keep an eye on Kakoune state, but if closed improperly can crash Kakoune

**Comment by mawww** [comment]
Could you describe the actual need you have, I think `WinClose` should go, but I'd like to make sure we provide a feature that matches your needs first.

**Comment by andreyorst** [comment]
> Could you describe the actual need you have, I think `WinClose` should go, but I'd like to make sure we provide a feature that matches your needs first.

I've wrote these [two commands](https://github.com/andreyorst/tagbar.kak/blob/eaacb915a86ffdc27f4f1488aefda652eadc4fdd/rc/tagbar.kak#L212-L238). After that I've aliased `:q` and `:wq` to them. This way if there are only two clients, and one of those is a `tagbarclient`, it will be closed before exiting Kakoune. 

I want a way to do this via hooks without making aliases that can break if something changes in `quit` or `write-quit` command, and it makes harder to use `:q!` and `:wq!` because now I need to write `:q !` and `:wq !` because if I will not add space before `!`, tagbar won't be closed, because it's another alias.

As I understand `WinClose` hook will not be run if I simply `:q` when another client exists, because window wasn't closed, but client was disconnected.
